plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse_token.go

```diff
@@ -0,0 +1,138 @@
+)
+
+// token represents the json structure that will be returned from the device after successfull authentication.
+// The token can be used for further API calls and needs to be encoded in the 'authorization' field of the
+// http header.
+type token struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+}
+
+// validToken respresents the json structure that will be returned by the device
+// after a request to check if the given token is still valid.
+type validToken struct {
+	Valid bool `json:"valid"`
+}
+
+// String builds the token string which is used in the 'authorization' field of the http header.
+func (t *token) String() string {
+	return t.TokenType + " " + t.AccessToken
+}
+
+type tokenManager struct {
+	url        string
+	username   string
+	password   string
+	connection *http.Client
+	log        telegraf.Logger
+
+	mux   sync.Mutex // mux used to serialize authorization requests
+	token token
+}
+
+// checkAuthToken checks if the tokenManager has valid authentication token.
+func (m *tokenManager) checkAuthToken() (isValid bool, err error) {
+	if m.token.AccessToken == "" {
+		return false, nil
+	}
+
+	u := m.url + validateTokenPath
+
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		return false, fmt.Errorf("failed to create request to validate token: %v", err)
+	}
+
+	req.Header.Add("Authorization", m.token.String())
+
+	resp, err := m.connection.Do(req)
+	if err != nil {
+		return false, fmt.Errorf("failed to execute request to validate token: %v", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return false, fmt.Errorf("failed to read response to validate token: %v", err)
+	}
+	var vt validToken
+	err = json.Unmarshal(body, &vt)
+	if err != nil {
+		return false, fmt.Errorf("failed to unmarshal validate token response: %v", err)
+	}
+	return vt.Valid, nil
+}
+
+// requestAuthToken gets a valid authentication token from the ctrlX CORE using username and password.
+// The token is necessary to create a subscription. An error will be returned, when no token could get aquired.
+func (m *tokenManager) requestAuthToken() (token, error) {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+
+	v, err := m.checkAuthToken()
+	if err != nil {
+		// Something went wrong during token validation
+		m.log.Error(err)
+	}
+	if v {
+		// Token was already created and token still valid
+		return m.token, nil
+	}
+
+	m.token = token{}
+
+	tokenURL := m.url + identityPath
+	payload, err := json.Marshal(map[string]string{
+		"name":     m.username,
+		"password": m.password,
+	})
+	if err != nil {
+		return m.token, fmt.Errorf("failed to create authorization token request: %v", err)
+	}
+	requestBody := bytes.NewBuffer(payload)
+
+	resp, err := m.connection.Post(tokenURL, "application/json", requestBody)
+	if err != nil {
+		r := strings.NewReplacer(identityPath, "/...")
+		b := r.Replace(err.Error())
```

Comment:
What is this doing and why?

Reply:
We want to shorten the error message and hide the api information to the end user.

Reply:
Can you please add your statement as a comment in the code!?